### PR TITLE
Add --all flag to ./x.py clean

### LINKED ISSUE #44214: `./x.py clean` does not clean LLVM, `./x.py clean -h` looks wrong
In a clone of this repository that I hadn’t used for a long time, I ran `git pull` then `./x.py build`. The latter failed in the middle of building LLVM with:

```
[ 15%] Linking CXX executable ../../bin/llvm-config
/usr/bin/ld: cannot find -lLLVM
collect2: error: ld returned 1 exit status
[ 15%] Linking CXX executable ../../bin/llvm-cxxfilt
tools/llvm-config/CMakeFiles/llvm-config.dir/build.make:95: recipe for target 'bin/llvm-config' failed
make[2]: *** [bin/llvm-config] Error 1
CMakeFiles/Makefile2:11680: recipe for target 'tools/llvm-config/CMakeFiles/llvm-config.dir/all' failed
make[1]: *** [tools/llvm-config/CMakeFiles/llvm-config.dir/all] Error 2
```

Running `./x.py clean` and then `./x.py build` again did not help. It turns out that `./x.py clean` does not clean `build/x86_64-unknown-linux-gnu/llvm/build`. Removing everything with `rm -r build/*` "fixed" the issue.

I tried `./x.py clean -h` to see if there’s an option to clean everything, but the output of that is confusing. It looks like the help message for `./x.py` without any subcommand.

```
Updating submodules
    Finished dev [unoptimized] target(s) in 0.0 secs
Usage: x.py <subcommand> [options] [<paths>...]

Subcommands:
    build       Compile either the compiler or libraries
    test        Build and run some test suites
    bench       Build and run some benchmarks
    doc         Build documentation
    clean       Clean out build directories
    dist        Build distribution artifacts
    install     Install distribution artifacts

To learn more about a subcommand, run `./x.py <subcommand> -h`

Options:
    -v, --verbose       use verbose output (-vv for very verbose)
    -i, --incremental   use incremental compilation
        --config FILE   TOML configuration file for build
        --build BUILD   build target of the stage0 compiler
        --host HOST     host targets to build
        --target TARGET target targets to build
        --on-fail CMD   command to run on failure
        --stage N       stage to build
        --keep-stage N  stage to keep without recompiling
        --src DIR       path to the root of the rust checkout
    -j, --jobs JOBS     number of jobs to run in parallel
    -h, --help          print this help message

Run `./x.py clean -h -v` to see a list of available paths.
```

### PR DESCRIPTION
**Pull request by tommyip**
This make `clean` removes the LLVM and download cache directory as well.

Fixes #44214.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by oli-obk** [comment]
Can we have a less nuclear option, too? After a rebase I frequently need to clean, but I really want to keep llvm.

**Comment by petrochenkov** [comment]
IIRC, there were `clean` and `clean-all` removing everything-except-for-LLVM and everything, respectively.

**Comment by Mark-Simulacrum** [comment]
@tommyip Would you be up to adding an additional subcommand to bootstrap (clean-all) that removes `build` and `tmp` and then leaving the existing clean as-is? I think that would satisfy all parties involved.

Such a step would need to be taken in `src/bootstrap/flags`, probably threading through into `config` and maybe a few other places -- I can provide more detailed instructions if you want them; let me know!

**Comment by tommyip** [comment]
Sure I will have a go at that tomorrow.

**Comment by tommyip** [comment]
Do we want `clean --all` or just `clean-all`?

**Comment by Mark-Simulacrum** [comment]
I'm fine with either. Can you also update the documentation for clean? It won't take target/host etc anymore.

**Comment by tommyip** [comment]
Where is that documentation?

File: src/bootstrap/flags.rs [hunk 1]
```diff
@@ -259,8 +262,10 @@ Arguments:
             let maybe_rules_help = Builder::get_help(&build, subcommand.as_str());
             extra_help.push_str(maybe_rules_help.unwrap_or_default().as_str());
         } else {
-            extra_help.push_str(format!("Run `./x.py {} -h -v` to see a list of available paths.",
-                     subcommand).as_str());
+            if subcommand.as_str() != "clean" {
+                extra_help.push_str(format!("Run `./x.py {} -h -v` to see a list of available paths.",
```
Comment (kennytm) [inline src/bootstrap/flags.rs]: This line is too long. CI rejects lines with >100 chars. Perhaps you should change that `else { if ... { ... } }` into `else if ... { ... }`.

```
[00:03:16] tidy error: /checkout/src/bootstrap/flags.rs:266: line longer than 100 chars
```

**Comment by tommyip** [comment]
@Mark-Simulacrum This PR is ready for review.

**Comment by Mark-Simulacrum** [comment]
Thanks!

@bors r+ rollup

**Comment by Mark-Simulacrum** [comment]
Merged in https://github.com/rust-lang/rust/pull/44936.

**Pull request closed without merging**